A14-channelz.md

```diff
@@ -0,0 +1,1051 @@
+
+### Server Channel Trace
+
+Servers keep track of recent interesting events that have happened.  This
+information is stored in the server channel trace.  The trace includes events
+such as connecting clients, timeouts, protocol violations, etc.  It also
+includes information about the shutdown of the server.  This is important as the
+server may have many active calls when it is shutting down, which prevent it
+from terminating.
+
+### Sockets and Listen Sockets
+
+Unlike channels, servers have two kinds of sockets.  A regular socket is one
+that is connected to a remote endpoint.  This is the most common type of socket.
+Servers also have a "listen" socket, which is used to accept and create new
+sockets.
+
+There may be no listen sockets, but active regular sockets.  This can occur
+during shutdown, when the server no longer wishes to accept connections but
+would like to gracefully complete the existing ones.  If a socket could exist in
+both locations, it may.  Implementations should prefer to segregate the two
+sockets.
+
+
+## Sockets
+
+Conceptually, sockets are the equivalent of a file descriptor.  They have a
+local and remote address, as well as some concept of security detail.  Sockets
+keep track of "streams" while Channels and Servers keep track of "calls."
+Sockets also expose access to underlying details of the operating system, such
+as socket options.  They are the lowest level in the channelz hierarchy.
+
+A SocketRef is associated with each socket, and contains a unique id and an
+optional name.  The id is unique among even the channels and servers.  The name
+includes information about the socket in a human readable form.  This is
+important as it is used as a "bread crumb" when deciding which socket to query
+for, from the context of a channel or server.
+
+### Socket Data
+
+The data exposed by a socket is much lower level than the channel or server
+data.  This information is meant to aid in debugging slow or stalled
+connections.  By default it includes the following:
+
+* Number of streams started, succeeded, or failed.
+* Number of gRPC messages sent and received.
+* Number of keepalive messages sent.
+* The last time a outbound stream was started and the last time an inbound
+  stream was started.
+* The last time a outbound message was sent and the last time an inbound
+  message was received.
+* The amount of flow control window granted to the remote endpoint, and the
+  amount of flow control window granted by remote endpoint.
+* All known socket options set.
+
+The stream counts differ slightly than that of the number of calls.  Calls may
+be started before a socket (transport) is available, so the stream counter may
+smaller than the corresponding call counters.  Conversely, if a stream fails and
+the call is retriable, a new stream may be started.  This could make the stream
+counts higher than the call counts.  Streams are considered successful if they
```

Comment:
This implies that even if the grpc call fails, the stream will be considered successful if trailing metadata containing the grpc status is sent successfully?

Reply:
Makes sense.  Might be a good idea to explicitly say that.